From cd381686fbe41b947c604058275174b602051b2f Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Sun, 22 Jun 2025 13:23:24 +0200
Subject: [PATCH 201/985] net: dovetail: fix release of in-band skb with oob
 storage

__kfree_skb() may receive in-band buffers (!skb_is_oob()) conveying
oob storage, which must be passed to the companion core for
release. Typically, this may happen when network interface taps are
fed with traffic handled by such core.

However, we want to allow the companion core to resend such buffers to
__kfree_skb() via finalize_skb_inband() once the storage is dropped
eventually, in order to flush their head state, then release the
buffer shells to the proper kmem cache.

Add the necessary logic to enable this path, using the new
oob_released marker to prevent a circular call chain to take place
(i.e. __kfree_skb -> skb_release_oob -> free_skb_oob ->
finalize_skb_inband -> __kfree_skb). Meanwhile, move the oob marker
out of the skb 'headers' group, since we don't want a clone to inherit
this information from the source skb, which would cause a release to
the wrong (oob) pool.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/linux/skbuff.h | 35 ++++++++++++++++++-----------------
 net/core/dev.c         |  2 +-
 net/core/skbuff.c      | 23 +++++++++++------------
 3 files changed, 30 insertions(+), 30 deletions(-)

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 245e35782..d0ee34964 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -949,6 +949,10 @@ struct sk_buff {
 				head_frag:1,
 				pfmemalloc:1,
 				pp_recycle:1; /* page_pool recycle indicator */
+#ifdef CONFIG_NET_OOB
+	__u8			oob:1,
+				oob_released:1;
+#endif
 #ifdef CONFIG_SKB_EXTENSIONS
 	__u8			active_extensions;
 #endif
@@ -995,9 +999,6 @@ struct sk_buff {
 	__u8			ndisc_nodetype:2;
 #endif
 
-#ifdef CONFIG_NET_OOB
-	__u8			oob:1;
-#endif
 #if IS_ENABLED(CONFIG_IP_VS)
 	__u8			ipvs_property:1;
 #endif
@@ -5285,6 +5286,16 @@ static inline void skb_mark_oob(struct sk_buff *skb)
 	skb->oob = 1;
 }
 
+static inline bool skb_is_oob_released(const struct sk_buff *skb)
+{
+	return skb->oob_released;
+}
+
+static inline void skb_mark_oob_released(struct sk_buff *skb)
+{
+	skb->oob_released = 1;
+}
+
 /**
  * skb_is_oob_timestamped - Whether the buffer is timestamped by a
  * companion core, in which case the timestamps should be stored in
@@ -5341,7 +5352,7 @@ static inline dma_addr_t skb_oob_dma_addr(const struct sk_buff *skb)
 	return page_pool_get_dma_addr(page);
 }
 
-bool recycle_skb_oob(struct sk_buff *skb);
+bool skb_release_oob(struct sk_buff *skb);
 void free_skb_oob(struct sk_buff *skb);
 
 static inline void finalize_skb_inband(struct sk_buff *skb)
@@ -5349,17 +5360,6 @@ static inline void finalize_skb_inband(struct sk_buff *skb)
 	__kfree_skb(skb);
 }
 
-/**
- *	__skb_inband_clone - In-band specific setup for skbs for a
- *	newly allocated clone head. Other allocation places don't need
- *	that because they are memset-zeroed from skb to &skb->end
- *	before building the buffer around a data storage area.
- */
-static inline void __skb_inband_clone(struct sk_buff *skb)
-{
-	skb->oob = 0;
-}
-
 /**
  *	__skb_oob_free_head - Called from the in-band net core after
  *      the last reference to the buffer (->users) was dropped, and
@@ -5385,8 +5385,9 @@ static inline bool skb_is_oob(const struct sk_buff *skb)
 	return false;
 }
 
-static inline void __skb_inband_clone(struct sk_buff *skb)
+static inline bool skb_is_oob_released(const struct sk_buff *skb)
 {
+	return false;
 }
 
 static inline bool __skb_oob_free_head(struct sk_buff *skb)
@@ -5414,7 +5415,7 @@ static inline dma_addr_t skb_oob_dma_addr(const struct sk_buff *skb)
 	return DMA_MAPPING_ERROR;
 }
 
-static inline bool recycle_skb_oob(struct sk_buff *skb)
+static inline bool skb_release_oob(struct sk_buff *skb)
 {
 	return false;
 }
diff --git a/net/core/dev.c b/net/core/dev.c
index 0e238ab83..abd435945 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3265,7 +3265,7 @@ void dev_kfree_skb_irq_reason(struct sk_buff *skb, enum skb_drop_reason reason)
 	}
 	get_kfree_skb_cb(skb)->reason = reason;
 
-	if (recycle_skb_oob(skb))
+	if (skb_release_oob(skb))
 		return;
 
 	local_irq_save(flags);
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 0a96c36f2..9cbeb8696 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -483,7 +483,7 @@ struct sk_buff *__build_skb(void *data, unsigned int frag_size)
 
 	if (running_oob()) {
 		skb = get_oob_skb();
-		if (unlikely(!skb))
+		if (unlikely(WARN_ON_ONCE(!skb)))
 			return NULL;
 	} else {
 		skb = kmem_cache_alloc(net_hotdata.skbuff_cache,
@@ -597,14 +597,15 @@ EXPORT_SYMBOL(sysctl_max_oob_skb);
 __weak void free_skb_oob(struct sk_buff *skb)
 { }
 
-bool recycle_skb_oob(struct sk_buff *skb)
+bool skb_release_oob(struct sk_buff *skb)
 {
 	/*
-	 * Hand the buffer release over the out-of-band core either if
-	 * the latter manages the data storage or we are currently
-	 * running oob.
+	 * If the buffer was not already released by the out-of-band
+	 * core, hand it the buffer for release either if it manages
+	 * the data storage or we are currently running oob.
 	 */
-	if (running_oob() || skb_is_oob_managed(skb)) {
+	if (!skb_is_oob_released(skb) && (running_oob() || skb_is_oob_managed(skb))) {
+		skb_mark_oob_released(skb);
 		free_skb_oob(skb);
 		return true;
 	}
@@ -1334,7 +1335,7 @@ static void skb_release_all(struct sk_buff *skb, enum skb_drop_reason reason)
 
 void __kfree_skb(struct sk_buff *skb)
 {
-	if (recycle_skb_oob(skb))
+	if (skb_release_oob(skb))
 		return;
 
 	skb_release_all(skb, SKB_DROP_REASON_NOT_SPECIFIED);
@@ -1619,7 +1620,7 @@ static void napi_skb_cache_put(struct sk_buff *skb)
 
 void __napi_kfree_skb(struct sk_buff *skb, enum skb_drop_reason reason)
 {
-	if (recycle_skb_oob(skb))
+	if (skb_release_oob(skb))
 		return;
 
 	skb_release_all(skb, reason);
@@ -1661,7 +1662,7 @@ void napi_consume_skb(struct sk_buff *skb, int budget)
 		return;
 	}
 
-	if (recycle_skb_oob(skb))
+	if (skb_release_oob(skb))
 		return;
 
 	skb_release_all(skb, SKB_CONSUMED);
@@ -2237,8 +2238,6 @@ struct sk_buff *skb_clone(struct sk_buff *skb, gfp_t gfp_mask)
 		n->fclone = SKB_FCLONE_UNAVAILABLE;
 	}
 
-	__skb_inband_clone(n);
-
 	return __skb_clone(n, skb);
 }
 EXPORT_SYMBOL(skb_clone);
@@ -6147,7 +6146,7 @@ EXPORT_SYMBOL(__skb_warn_lro_forwarding);
 void kfree_skb_partial(struct sk_buff *skb, bool head_stolen)
 {
 	if (head_stolen) {
-		if (!recycle_skb_oob(skb)) {
+		if (!skb_release_oob(skb)) {
 			skb_release_head_state(skb);
 			if (!__skb_oob_free_head(skb))
 				kmem_cache_free(net_hotdata.skbuff_cache, skb);
-- 
2.30.2

